query/src/cmd/compaction.rs

```diff
@@ -0,0 +1,255 @@
+            if let Some(event) = resp.event {
+                if event.prev.is_none() || event.current.is_none() {
+                    continue;
+                }
+
+                let key = event.key;
+                let id = TableId::from_key(&key).unwrap();
+
+                let current_seqv = event.current.unwrap();
+                let seq = current_seqv.seq;
+                let meta: TableMeta = deserialize_struct(&current_seqv.data).unwrap();
+                if meta.drop_on.is_some() || meta.default_cluster_key.is_some() {
+                    continue;
+                }
+
+                let ident = TableIdent::new(id.table_id, seq);
+                let table_info = TableInfo {
+                    ident,
+                    desc: "".to_owned(),
+                    name: "".to_owned(),
+                    meta,
+                };
+                delay_queue_insert
+                    .insert(id.table_id, table_info)
+                    .await
+                    .unwrap();
+            }
+        }
+    });
+
+    let catalog = ctx.get_catalog(CATALOG_DEFAULT)?;
+    while let Some(info) = delay_queue_read.next().await {
+        let table = catalog.get_table_by_info(&info)?;
+        let ctx = Arc::clone(&ctx);
+        ctx.get_storage_runtime()
+            .spawn(async move { do_compaction(ctx.clone(), table, "".to_owned()).await });
+    }
+
+    Ok(())
+}
+
+pub async fn do_compaction(
+    ctx: Arc<QueryContext>,
+    table: Arc<dyn Table>,
+    database: String,
+) -> common_exception::Result<()> {
+    if !table.auto_compaction() {
+        return Ok(());
+    }
+
+    let plan = OptimizeTablePlan {
+        catalog: ctx.get_current_catalog(),
+        database,
+        table: table.name().to_string(),
+        action: OptimizeTableAction::Compact,
+    };
+
+    table.compact(ctx.clone(), plan).await
+}
+
+struct CompactionQueue {
+    timeout: Duration,
+
+    entries: HashMap<u64, (TableInfo, delay_queue::Key)>,
+    expirations: DelayQueue<u64>,
+}
+
+impl CompactionQueue {
+    fn new(timeout: Duration) -> Self {
+        Self {
+            timeout,
+            entries: HashMap::new(),
+            expirations: DelayQueue::new(),
+        }
+    }
+
+    fn insert(&mut self, table_id: u64, table_info: TableInfo) {
+        if let Some((_, key)) = self.entries.get(&table_id) {
+            self.expirations.reset(key, self.timeout);
+            self.entries.insert(table_id, (table_info, *key));
+        } else {
+            let key = self.expirations.insert(table_id, self.timeout);
+            self.entries.insert(table_id, (table_info, key));
+        }
+    }
+
+    fn poll_expired(&mut self, cx: &mut Context) -> Poll<Option<TableInfo>> {
+        match ready!(self.expirations.poll_expired(cx)) {
+            Some(key) => {
+                let table_info = self.entries.remove(key.get_ref()).unwrap().0;
+                Poll::Ready(Some(table_info))
+            }
+            None => Poll::Pending,
```

Comment:
I might be wrong, but will cpu be trapped in a busy loop of polling here if `DealyQueue` is empty?

Since `self.expirations.poll_expired(cx)` keeps returning `None`, if the DelayQueue is empty, and here we just return `Pending` to the puller.